storage/segment_query.go

```diff
@@ -484,39 +505,42 @@ func collectSingleFieldGroupByResults(
 
 		// Add values to calculation results.
 		var (
-			emptyValueUnion calculation.ValueUnion
-			calcFieldIdx    int
+			emptyValueUnion   calculation.ValueUnion
+			calcFieldIdx      int
+			calcVals, hasVals = calcIter.Values() // len(calcVals) == number of calc fields
 		)
 		for i, calc := range res.Calculations {
 			if !calc.Op.RequiresField() {
 				calcResults[i].Add(emptyValueUnion)
-			} else {
-				cv := toCalcValueFns[calcFieldIdx](&calcVals[calcFieldIdx])
-				calcResults[i].Add(cv)
-				calcFieldIdx++
+			} else if !hasVals[calcFieldIdx] {
+				// The calculation field does not have value for this iteration.
```

Comment:
you're going to have to explain this one to me offline, it looks like we want to do something like:
```
for i, calc := range res.Calculations {
    for calcFieldIdx, calcFieldIdx < len(calcVals); calcFieldIdx++ {
       // run calc for every value that exists
    }
}
```